source/mir/math/stat.d

```diff
@@ -420,6 +424,357 @@ unittest
+{
+    mixin("alias average = .average!(F, CentralTendency." ~ centralTendency ~ ", Summation." ~ summation ~ ");");
+}
+
+///ditto
+template average(string centralTendency, string summation = "appropriate")
+{
+    mixin("alias average = .average!(CentralTendency." ~ centralTendency ~ ", Summation." ~ summation ~ ");");
+}
+
+/// Generic average of vector
+version(mir_test)
+@safe pure nothrow @nogc
+unittest
+{
+    import mir.math.common: approxEqual;
+
+    static immutable x = [20.0, 100.0, 2000.0, 10.0, 5.0, 2.0];
+    
+    assert(x.average!"mean".approxEqual(356.1667));
+    assert(x.average!"hmean".approxEqual(6.97269));
+}
+
+/// Generic average of matrix
+version(mir_test)
+pure @safe
+unittest
+{
+    import mir.math.common: approxEqual;
+    import mir.ndslice.fuse: fuse;
+
+    auto x = [
+        [20.0, 100.0, 2000.0], 
+        [10.0,   5.0,    2.0]
+    ].fuse;
+
+    assert(x.average!"mean".approxEqual(356.1667));
+    assert(x.average!"hmean".approxEqual(6.97269));
+}
+
+/// Column generic average of matrix
+version(mir_test)
+pure @safe
+unittest
+{
+    import mir.algorithm.iteration: all;
+    import mir.math.common: approxEqual;
+    import mir.ndslice: fuse;
+    import mir.ndslice.topology: alongDim, byDim, map;
+
+    auto x = [
+        [20.0, 100.0, 2000.0],
+        [10.0,   5.0,    2.0]
+    ].fuse;
+
+    auto y_mean = [15.0, 52.5, 1001];
+    auto y_hmean = [13.33333, 9.52381, 3.996004];
+
+    // Use byDim or alongDim with map to compute average of row/column.
+    assert(x.byDim!1.map!(average!"mean").all!approxEqual(y_mean));
+    assert(x.alongDim!0.map!(average!"mean").all!approxEqual(y_mean));
+    
+    assert(x.byDim!1.map!(average!"hmean").all!approxEqual(y_hmean));
+    assert(x.alongDim!0.map!(average!"hmean").all!approxEqual(y_hmean));
+}
+
+/// Can also pass arguments to average
+version(mir_test)
+pure @safe nothrow @nogc
+unittest
+{
+    import mir.ndslice.topology: repeat;
+    import mir.math.common: approxEqual;
+
+    //Set sum algorithm or output type
+    static immutable x = [1, 1e-100, 1, -1e-100];
+
+    assert(x.average!("mean", "kb2").approxEqual(0.5));
+    assert(x.average!("mean", "precise").approxEqual(0.5));
+    
+    assert(x.average!("hmean", "kb2").approxEqual(2));
+    assert(x.average!("hmean", "precise").approxEqual(2));
+
+    //Provide the summation type
+    assert(float.max.repeat(3).average!(double, "mean", "fast").approxEqual(float.max));
+    assert(float.max.repeat(3).average!(double, "hmean", "fast").approxEqual(float.max));
+}
+
+private template centerImpl(F, CentralTendency centralTendency, 
+                            Summation summation)
+{
+    @safe pure nothrow
+    auto centerImpl(Range)(Range r, F average)
+        if (isIterable!Range)
+    {
+        static if (__traits(compiles, r - average)) {
+            return r - average;
+        } else {
+            import mir.ndslice.topology: map;
+            return r.map!(a => a - average);
```

Comment:
it will allocate lambda in GC, also it will really never compiles because Mir's map works only with ndslices.